caffe2/onnx/backend.cc

```diff
@@ -0,0 +1,1080 @@
+#include "backend.h"
+#include "device.h"
+#include "helper.h"
+#include "caffe2/core/logging.h"
+#include "onnx/optimizer/optimize.h"
+#include "onnx/defs/schema.h"
+#include "onnx/checker.h"
+
+#include "google/protobuf/io/coded_stream.h"
+#include "google/protobuf/io/zero_copy_stream_impl_lite.h"
+
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <sstream>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace onnx_caffe2 {
+
+namespace detail {
+
+constexpr static int kKnownOpsetVersion = 3;
+
+bool AlmostEqual(double a, double b) {
+  constexpr static double kEps = 1e-15;
+  return (fabs(a -  b) < kEps);
+}
+
+template <class T>
+bool TryConvertingTensorRawValues(const onnx::TensorProto &onnx_tensor,
+                                  ::google::protobuf::RepeatedField<T> *field) {
+  if (not onnx_tensor.has_raw_data()) {
+    return false;
+  }
+
+  size_t raw_size = onnx_tensor.raw_data().size();
+  assert(raw_size % sizeof(T) == 0);
+
+  size_t num_elements = raw_size / sizeof(T);
+  const void *src_ptr =
+      static_cast<const void *>(onnx_tensor.raw_data().data());
+  field->Resize(num_elements, 0);
+  void *target_ptr = static_cast<void *>(field->mutable_data());
+  memcpy(target_ptr, src_ptr, raw_size);
+
+  return true;
+}
+
+//TODO We probably should have only one copy of this function (copied from pybind_state.cc)
+bool ParseProtobufFromLargeString(const std::string &str, ::google::protobuf::Message *proto) {
+  ::google::protobuf::io::ArrayInputStream input_stream(str.data(), str.size());
+  ::google::protobuf::io::CodedInputStream coded_stream(&input_stream);
+  // Set PlanDef message size limit to 1G.
+  coded_stream.SetTotalBytesLimit(1024LL << 20, 512LL << 20);
+  return proto->ParseFromCodedStream(&coded_stream);
+}
+
+caffe2::DeviceOption GetDeviceOption(const onnx::Device &onnx_device) {
+  static const std::unordered_map<onnx::DeviceType, caffe2::DeviceType> m = {
+    {onnx::DeviceType::CPU, caffe2::DeviceType::CPU},
+    {onnx::DeviceType::CUDA, caffe2::DeviceType::CUDA}
+  };
+  caffe2::DeviceOption d;
+  d.set_device_type(static_cast<int32_t>(m.at(onnx_device.type)));
+  d.set_cuda_gpu_id(onnx_device.device_id);
+  return d;
+}
+
+// Note: if we design the API a bit, we can avoid the two copies
+onnx::ModelProto OptimizeOnnx(const onnx::ModelProto &input, bool init) {
+  std::vector<std::string> passes{"fuse_consecutive_transposes",
+                                  "eliminate_nop_transpose",
+                                  "fuse_transpose_into_gemm"};
+
+  if (init) {
+    passes.emplace_back("split_init");
+  } else {
+    passes.emplace_back("split_predict");
+  }
+  return onnx::optimization::Optimize(input, passes);
+}
+
+template <class T, class U>
+U LookUpWithDefault(const std::unordered_map<T, U> &map, const T &key,
```

Comment:
nit: you can follow folly's convention and call it `get_default` (not that it matters much)

Reply:
Maybe put it in some common Caffe2 util file?